File: nodejs/consumer.js
const opentracing = require("opentracing")
const mysql = require('mysql2/promise')
const boot = require('./boot').boot

async function main() {
    const [tracer, channel, appSpan] = await boot('consumer')

    const connectingSpan = tracer.startSpan('mysql-connecting', {childOf: appSpan})
    const pool = await initConnectionsPool()
    connectingSpan.finish()

    const consumingSpan = tracer.startSpan('consuming', {childOf: appSpan})

    process.on('SIGINT', async function () {
        console.log('flushing')
        consumingSpan.finish()
        appSpan.finish()
        await tracer.close()
        // process.exit()
    });

    let counter = 0
    channel.consume('message_queue', async function (message) {
        counter++
        const parentSpan = tracer.extract(opentracing.FORMAT_HTTP_HEADERS, message.properties.headers)
        const span = tracer.startSpan(message.fields.routingKey, {childOf: parentSpan})

        await processMessage(message, pool)

        await publishEvent(message, channel, tracer, span)

        channel.ack(message)

        span.finish()
    })

    async function processMessage(message, pool) {
        const msg = JSON.parse(message.content)
        const conn = await pool.getConnection()
        await conn.query('START TRANSACTION')
        const [rows] = await conn.query('SELECT value FROM data WHERE id=?', [msg.id])
        const currentValue = rows[0]['value']
        await conn.query('UPDATE data SET value=? WHERE id=?', [currentValue + msg.value, msg.id])
        await conn.query('COMMIT')
        conn.release()
    }

    async function publishEvent(message, channel, tracer, span) {
        const headers = []
        tracer.inject(span, opentracing.FORMAT_HTTP_HEADERS, headers)
        await channel.sendToQueue('events_queue', message.content, {headers: headers})
    }

    function initConnectionsPool() {
        return mysql.createPool({
            host: process.env.DB_HOST,
            database: process.env.DB_NAME,
            user: process.env.DB_USER,
            password: process.env.DB_PASS,
            waitForConnections: true,
            connectionLimit: process.env.DB_POOL,
            queueLimit: 0
        })
    }
}

main()

File: golang/consumer.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"strconv"

	_ "github.com/go-sql-driver/mysql"
	"github.com/joho/godotenv"
	"github.com/streadway/amqp"
)

func main() {
	godotenv.Load("../.env")

	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s)/%s", os.Getenv("DB_USER"), os.Getenv("DB_PASS"), os.Getenv("DB_HOST"), os.Getenv("DB_NAME")))
	if err != nil {
		fmt.Println("Failed connecting MySQL")
		panic(err)
	}
	poolSize, _ := strconv.Atoi(os.Getenv("DB_POOL"))
	db.SetMaxOpenConns(poolSize)

	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
	if err != nil {
		fmt.Println("Failed connecting RabbitMQ")
		panic(err)
	}

	ch, err := conn.Channel()
	if err != nil {
		fmt.Println(err)
	}
	defer ch.Close()

	if err != nil {
		fmt.Println(err)
	}

	msgs, err := ch.Consume(
		"message_queue",
		"",
		true,
		false,
		false,
		false,
		nil,
	)

	forever := make(chan bool)
	go func() {
		for d := range msgs {
			// fmt.Printf("Processing: %s\n", d.Body)
			processMessage(db)
			publishEvent(chan)
		}
	}()

	fmt.Println("Consuming")
	<-forever
}

func processMessage(db) {
    var msg map[string]interface{}
    json.Unmarshal([]byte(d.Body), &msg)
    db.Exec("START TRANSACTION")
    db.Exec("SELECT value FROM data WHERE id=?", msg["id"])
    db.Exec("UPDATE data SET value=? WHERE id=?", msg["value"], msg["id"])
    db.Exec("COMMIT")
}

func publishEvent() {

}